posix/netdb.c

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/corelibc/netdb.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/sys/socket.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/posix/resolver.h>
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/trace.h>
+
+static size_t _resolver_table_len = 3;
+static oe_resolver_t* _resolver_table[3] = {0}; // At most 3
+static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
+
+/* Called by the public oe_load_module_host_resolver() function. */
+int oe_register_resolver(int resolver_priority, oe_resolver_t* presolver)
+{
+    int ret = -1;
+
+    oe_spin_lock(&_lock);
+
+    if (presolver == NULL)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (resolver_priority >= (int)_resolver_table_len)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    _resolver_table[resolver_priority] = presolver;
+
+    ret = 0;
+done:
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+int oe_getaddrinfo(
+    const char* node,
+    const char* service,
+    const struct oe_addrinfo* hints,
+    struct oe_addrinfo** res_out)
+{
+    int ret = OE_EAI_FAIL;
+    size_t i;
+
+    oe_spin_lock(&_lock);
+
+    if (res_out)
```

Comment:
Optional: this can be outside the lock, it introduces no branching.

Reply:
I have tried to use the **locked** boolean where possible. Sometimes the locking region is bigger than needed to reduce extra checking. Perhaps we need a lock type with a built-in **locked** variable.